File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<Integer> output = new ArrayList<>();
        List<List<Integer>> ans = new ArrayList<>();
        solve(nums,0,output,ans);
        return ans;
    }
    
    public void solve(int[] nums, int i, List<Integer> output, List<List<Integer>> ans){
        
        if(i == nums.length){
            ans.add(new ArrayList<Integer>(output));
            return;
        }
        
        output.add(nums[i]);
        solve(nums,i+1,output,ans);
        output.remove(output.size()-1);
        solve(nums,i+1,output, ans);
    }
}

File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        def backtrack(start, path):
            ans.append(path[:])
            for i in range(start, len(nums)):
                path.append(nums[i])
                backtrack(i+1, path)
                path.pop()
        backtrack(0,[])
        return ans
        